cmake: make library discoverable via find_package()

hgkjshegfskef opened the pull request:
Library can be discovered through find_package(CONFIG) mode. Library
produces Trace::trace IMPORTED target to be linked against. The target
is discoverable both in installed state, and when the project is
included as subdirectory.

Additionally, a few minor cleanups are made:
- no need to look for C compiler, there are no C files
- no need to print information about C compiler (no need to print
  information about C++ compiler either, since this is done by CMake
  automatically, but since it exists already, it is probably for a
  reason)
- library version is specified (required for target versioning)
- enable_testing() is wrapped inside BUILD_TESTING, which is a standard
  CMake variable, and set to ON by default to preserve existing behavior

After installation (tested) or inclusion as subdirectory (not tested),
expected usage is:

    find_package(Trace CONFIG REQUIRED)
    target_link_libraries(foobar PRIVATE Trace::trace)

Library can be included via:

    #include <trace/trace.hpp>

hgkjshegfskef committed:
cmake: make library discoverable via find_package()

Library can be discovered through find_package(CONFIG) mode. Library
produces Trace::trace IMPORTED target to be linked against. The target
is discoverable both in installed state, and when the project is
included as subdirectory.

Additionally, a few minor cleanups are made:
- no need to look for C compiler, there are no C files
- no need to print information about C compiler (no need to print
  information about C++ compiler either, since this is done by CMake
  automatically, but since it exists already, it is probably for a
  reason)
- library version is specified (required for target versioning)
- enable_testing() is wrapped inside BUILD_TESTING, which is a standard
  CMake variable, and set to ON by default to preserve existing behavior

After installation (tested) or inclusion as subdirectory (not tested),
expected usage is:

    find_package(Trace CONFIG REQUIRED)
    target_link_libraries(foobar PRIVATE Trace::trace)

Library can be included via:

    #include <trace/trace.hpp>

GPMueller commented on CMakeLists.txt:
```diff
@@ -9,13 +9,12 @@ option( TRACE_TEST_COVERAGE "Build trace tests with coverage flags" OFF )
 
 ######### Project Name ####################################
-project( trace )
+project( Trace LANGUAGES CXX VERSION 0.1 )
```
Any reason to change the name to uppercase `Trace`?

hgkjshegfskef replied:
Yes, this is to use `${PROJECT_NAME}` without directly using "Trace" to refer to the namespace of the target, as in, `Trace::trace` (usually, target's namespace is capitilized in CMake code, but it is not a firmly established practice), so that perhaps somebody can look at the code and practically copy it verbatim, as the procedure of creating and exporting targets and configs is pretty much boilerplate. I considered changing project name to be acceptable, since target name `trace` is still lowercase.

GPMueller commented on CMakeLists.txt:
```diff
@@ -63,51 +62,102 @@ set( TRACE_SOURCE_FILES
     COMPILE_FLAGS         "${TRACE_COMPILE_FLAGS} ${TRACE_FLAGS_COVERAGE}"
-    LINK_FLAGS            "${TRACE_LINK_FLAGS}    ${TRACE_FLAGS_COVERAGE}" )
+    LINK_FLAGS            "${TRACE_LINK_FLAGS}    ${TRACE_FLAGS_COVERAGE}"
+    VERSION               "${${PROJECT_NAME}_VERSION}"
```
I believe `PROJECT_NAME` is set by the topmost `project(...)` command, so it would be safer to use `LIBRARY_NAME` here and in the following

hgkjshegfskef replied:
There is `PROJECT_NAME` and `CMAKE_PROJECT_NAME`. The former always refers to the innermost project, the latter always refers to the outermost project. That is, if somebody adds this library as subdirectory, CMAKE_PROJECT_NAME will refer to their project, and PROJECT_NAME will refer to this library.

GPMueller commented on CMakeLists.txt:
```diff
@@ -63,51 +62,102 @@ set( TRACE_SOURCE_FILES
+
+
+######### Tests ###########################################
+option( BUILD_TESTING ON )
```
Please place `option`s near the top of the CMakeLists.txt, so they are obvious when reading this file.

Something about your changes (I cannot currently see the reason) is also keeping the tests from being built on the CI, which is therefore currently failing.

hgkjshegfskef replied:
Right. I remember now: I if'ed all the tests because I couldn't compile `catch` due to https://github.com/catchorg/Catch2/issues/2178 . You should probably update the header.

GPMueller replied:
I created a corresponding issue #3

GPMueller reviewed (changes requested):
Thanks for your PR!

As I merged your PR #2, please remove your second commit (to remove the semicolons) from this PR.

If you can get the CI running again and apply the small changes I mentioned, I'd be happy to merge your PR.

hgkjshegfskef committed:
cmake: introduce TRACE_BUILD_TESTING, fix CI

TRACE_BUILD_TESTING is used to control whether tests are built. It is
off by default, in case downstream includes the library as subdirectory.

Additionally, clean up the testing section of cmake code:
- refer only to CURRENT variables instead of global ones, for the same
  reason as above
- no need to set WORKING_DIRECTORY on the test target

Additionally, fix CI build:
- at least on ubuntu-latest target, the dependencies were not met
- make bash shell the default shell for all jobs
- use single quotes around ${{runner.workspace}} to prevent bash from
  treating the '\' windows path separators as escapes (thanks Donovan
  Baarda on SO)
- "steps:" indentation matches github actions docs' examples

hgkjshegfskef commented:
Forgot to write, this is ready for another review.

GPMueller commented on .github/workflows/ci.yml:
```diff
@@ -18,13 +18,21 @@ jobs:
+    - name: 📚 Install necessary packages
+      shell: bash
+      run:  |
+            if [ "$RUNNER_OS" == "Linux" ]; then
```
Instead of using bash's if, please use the GH Actions way of doing it with `if: matrix.os == 'ubuntu-latest'`

GPMueller reviewed (changes requested):
Except the `if` in the CI script, everything looks very good 👍
We can ignore the minor coverage change perceived by codecov.

hgkjshegfskef commented:
New changes, fixed up to last commit:
- styled if check as requested
- removed if check on coverage job, since I realized it only runs on ubuntu
- added bash shell in 'create build dir' step for consistency with all other steps

Ready for another review.

hgkjshegfskef commented:
My bad, working-directory broke.

hgkjshegfskef commented:
Now the CI is not triggered at all. I don't get it.

hgkjshegfskef commented:
CI fixed.

Changes:
- make bash shell the default shell for all jobs
- use single quotes around ${{runner.workspace}} to prevent bash from
  treating the '\' windows path separators as escapes (thanks Donovan
  Baarda on SO)
- "steps:" indentation matches github actions docs' examples

Ready for review.

GPMueller commented on .github/workflows/ci.yml:
```diff
@@ -16,64 +24,66 @@ jobs:
 
-    - name: 📁 Create build folder
-      run: cmake -E make_directory ${{runner.workspace}}/build
+      - name: 📚 Install necessary packages
+        if: matrix.os == 'ubuntu-latest'
+        run: |
+          sudo apt update
+          sudo apt install -y cmake
```
why did you add an installation for cmake? I don't believe it's necessary

hgkjshegfskef replied:
Apparently, https://github.com/nektos/act is not configured to use github's runners by default, and I was using it to test CI locally, which was failing due to missing dependenecies. And apparently, github images ship with a ginormous database of preinstalled packages (13G). Therefore, you are correct, it is not necessary. However, I suggest to keep this in order to not have to have the github's giant image with all the packages stored and updated locally on disk, because local CI runs become a pain.

GPMueller replied:
hmm interesting consideration. I tend to be minimalist, i.e. to only install what I really need, but this is fine

GPMueller commented on .github/workflows/ci.yml:
```diff
@@ -16,64 +24,66 @@ jobs:
+      - name: 📚 Install necessary packages
+        run: |
+          sudo apt update
+          sudo apt install -y cmake python3-pip lcov
```
why did you add `python3-pip` to the installed packages? I don't think it's necessary

hgkjshegfskef replied:
Same as the above.
